File: src/compute/jl/vhat_sigmahat.jl
"""
Calculate covariance matrix of lag j, this assumes that the time index is over the columns.
Thus, y is a N x T matrix, where N is the number of variables and T is the number of time periods.

## Arguments
- `y::Matrix{Float64}`: matrix of shape (N, T) where N is the number of variables and T is the number of time periods.
- `j::int`: number of lags.
- `bias::bool=false`: whether to use bias correction in covariance calculation.

## Returns
- `Matrix{Float64}`: covariance matrix of lag j with shape (N, N).
"""
function calc_Σj(y::Matrix{Float64}, j::Int)::Matrix{Float64}
    p, T = size(y)
    sigma_j = zeros(p, p)
    ymean = mean(y, dims=2)
    for t in 1:(T-j)
        @inbounds sigma_j += (y[:, t] - ymean) * (y[:, t+j] - ymean)'
    end
    return sigma_j / T
end

"""
Return a banded matrix with bandwidth h on both sides of the main diagonal.

## Arguments
- `A::Matrix{Float64}`: The input matrix.
- `h::int`: The half-bandwidth to use.

## Returns
- `Matrix{Float64}`: A banded matrix with bandwidth h on both sides of the main diagonal.
"""
function band_matrix(A::Matrix{Float64}, h::Int)::Matrix{Float64}
    N, T = size(A)
    B = zeros(N, T)
    for i in 1:N
        for j in max((i - h), 1):min((i + h), T)
            @inbounds B[i, j] = A[i, j]
        end
    end
    return B
end

"""
Construct the V matrix containing the columns of C' = [A B]', that are nonzero.

## Arguments
- `Σ0::Matrix{Float64}`: The covariance matrix of y.
- `Σ1::Matrix{Float64}`: The covariance matrix of lag 1.
- `h::int=0`: The bandwidth to use, if 0, then no bandwidth is used.
- `prebanded::bool=false`: Whether Σ0 and Σ1 are already banded. Default is false.

## Returns
- `Matrix{Float64}`: The V matrix with dimensions N x (N+1), containing the columns of C' = [A B]', that are nonzero.
"""
function constr_Vhat(Σ0::Matrix{Float64}, Σ1::Matrix{Float64}, h0::Int=0, h1::Int=0)::Matrix{Float64}
    if h0 == 0 || h1 == 0
        return [Σ1' Σ0]
    end

    Σ0 = band_matrix(Σ0, h0)
    Σ1 = band_matrix(Σ1, h1)
    return [Σ1' Σ0]
end

"""
Construct the vectorized autocovariance of lag 1 from the banded autocovariance matrix.

## Arguments
- `Σ1::Matrix{Float64}`: The banded autocovariance matrix of lag 1.
- `h::Int=0`: The half-bandwidth to use. Default is size(Σ1, 1)/4.

## Returns
- `Vector{Float64}`: The vectorized autocovariance of lag 1.
"""
function vec_sigma_h(Σ1::Matrix{Float64}, h1::Int=0)::Vector{Float64}
    if h1 == 0
        return vec(Σ1')
    end
    Σ1 = band_matrix(Σ1, h1)
    return vec(Σ1')
end

"""
Calculate the active column indices for the matrix V_h^(d).

## Arguments
- `p::Int`: The dimension of the matrix V.
- `bandwidth::Int=0`: The bandwidth to use. Default is p/4.

## Returns
- `Vector{Vector{Bool}}`: A p-length vector containing Boolean vectors representing the active column indices for each row.
"""
function active_cols(p::Int, bandwidth::Int=0)::Vector{Vector{Bool}}
    if bandwidth == 0
        bandwidth = div(p, 4)
    end

    active_set = [zeros(Bool, p * 2) for _ in 1:p]

    for i in 1:p
        lower_a = collect(max(1, i - bandwidth):max(0, i - 1))
        upper_a = collect((i+1):min(i + bandwidth, p))
        full_b = collect((p+max(1, (i - bandwidth))):(p+min(i + bandwidth, p)))
        selection = vcat(lower_a, upper_a, full_b)

        setindex!(active_set[i], ones(Bool, length(selection)), selection)
    end
    return active_set
end

"""
Construct a diagonal block matrix from the active columns of V.

## Arguments
- `V::Matrix{Float64}`: The V matrix with dimensions N x (N+1), containing the columns of C' = [A B]', that are nonzero.
- `bandwidth::Int=0`: The half-bandwidth to use. Default is size(V, 1)/4.

## Returns
- `SparseMatrixCSC{Float64}`: The resulting diagonal block matrix with dimensions p^2 x K, where K is the number of nonzero columns in V.
"""
function constr_Vhat_d(V::Matrix{Float64}, bandwidth::Int=0)::SparseMatrixCSC{Float64}
    p = size(V, 1)
    active = active_cols(p, bandwidth)
    res = [V[:, active[i]] for i in 1:p]
    Vhat_d = spzeros(p^2, sum(sum(active)))

    row_index = col_index = 1
    for (M, act) in zip(res, active)
        @inbounds setindex!(Vhat_d, M, collect(row_index:(row_index+p-1)), collect(col_index:(col_index+sum(act)-1)))
        row_index += p
        col_index += sum(act)
    end
    return Vhat_d
end

File: src/compute/R/vhat_sigmahat.R
library(Matrix)

calc_Sigma_j <- function(y, j) {
    p <- nrow(y)
    T <- ncol(y)
    sigma_j <- matrix(0, p, p)
    ymean <- rowMeans(y)
    for (t in 1:(T - j)) {
        sigma_j <- sigma_j + (y[, t] - ymean) %*% t(y[, t + j] - ymean)
    }
    return(sigma_j / T)
}

band_matrix <- function(A, h) {
    N <- nrow(A)
    T <- ncol(A)
    B <- matrix(0, N, T)
    for (i in 1:N) {
        for (j in max((i - h), 1):min((i + h), T)) {
            B[i, j] <- A[i, j]
        }
    }
    return(B)
}

construct_Vhat <- function(Sigma0, Sigma1, h0 = 0, h1 = 0) {
    if (h0 == 0 || h1 == 0) {
        return(cbind(t(Sigma1), Sigma0))
    }
    Sigma0 <- band_matrix(Sigma0, h0)
    Sigma1 <- band_matrix(Sigma1, h1)
    return(cbind(t(Sigma1), Sigma0))
}

construct_sigma_hat <- function(Sigma1, h1 = 0) {
    if (h1 == 0) {
        return(as.vector(t(Sigma1)))
    }
    res <- band_matrix(Sigma1, h1)
    return(as.vector(t(res)))
}

active_cols <- function(p, bandwidth = 0) {
    if (bandwidth == 0) {
        bandwidth <- floor((p - 1) / 4)
    }

    active_set <- matrix(FALSE, p, p * 2)

    for (i in 1:p) {
        full_a <- max(1, i - bandwidth):min(i + bandwidth, p)
        full_b <- (p + max(1, (i - bandwidth))):(p + min(i + bandwidth, p))
        selection <- c(full_a, full_b)

        active_set[i, selection] <- TRUE

        # Diagonal elements of A are set to FALSE
        active_set[i, i] <- FALSE
    }
    return(active_set)
}

active_cols_alt <- function(p, m) {
    active_set_A <- matrix(FALSE, p, p * 2)
    active_set_B <- matrix(FALSE, p, p * 2)

    for (i in 1:p) {
        for (j in 1:p) {
            # Vertical neighbours
            if (abs(i - j) == m) {
                active_set_A[i, j] <- TRUE
                active_set_B[i, j] <- TRUE
            }

            # Conditions
            not_side_edge <- !((i %% m == 0 && (j - 1) %% m == 0) || ((i - 1) %% m == 0 && j %% m == 0))
            not_top_edge <- (i > m) && (j > m)
            not_bottom_edge <- (i <= (p - m)) && (j <= (m * (m - 1)))

            # Horizontal neighbours
            if (abs(i - j) == 1 && not_side_edge) {
                active_set_A[i, j] <- TRUE
                active_set_B[i, j] <- TRUE
            }

            # Bottom right diagonal neighbour
            if (not_side_edge && not_bottom_edge && abs(i - j) == (m + 1)) {
                active_set_A[i, j] <- TRUE
                active_set_B[i, j] <- TRUE
            }

            # Top right diagonal neighbour
            if (not_side_edge && not_top_edge && abs(i - j) == (m - 1)) {
                active_set_A[i, j] <- TRUE
                active_set_B[i, j] <- TRUE
            }

            # Bottom left diagonal neighbour
            if (not_bottom_edge && not_side_edge && abs(i - j) == (m - 1)) {
                active_set_A[i, j] <- TRUE
                active_set_B[i, j] <- TRUE
            }

            # Top left diagonal neighbour
            if (not_top_edge && not_side_edge && abs(i - j) == (m + 1)) {
                active_set_A[i, j] <- TRUE
                active_set_B[i, j] <- TRUE
            }

            # Zero diagonal for A
            if (i == j) {
                active_set_A[i, j] <- FALSE
            }
        }
    }
    active_set <- cbind(active_set_A, active_set_B)

    return(active_set)
}

construct_Vhat_d <- function(V, bandwidth = 0) {
    p <- nrow(V)
    active <- active_cols(p, bandwidth)
    Vhat_d <- Matrix(0, p^2, sum(active), sparse = TRUE)

    row_index <- col_index <- idx <- 1
    for (idx in 1:p) {
        M <- V[, active[idx, ]]
        Vhat_d[row_index:(row_index + p - 1), col_index:(col_index + dim(M)[2] - 1)] <- M
        row_index <- row_index + p
        col_index <- col_index + dim(M)[2]
    }
    return(Vhat_d)
}
